scene/3d/area_3d.cpp

```diff
@@ -271,6 +276,7 @@ void Area3D::_body_inout(int p_status, const RID &p_body, ObjectID p_instance, i
 
 		if (!node || E->value.in_tree) {
 			emit_signal(SceneStringNames::get_singleton()->body_shape_entered, p_body, node, p_body_shape, p_area_shape);
+			node->emit_signal(SceneStringNames::get_singleton()->entered_area_shape, get_rid(), this, p_area_shape, p_body_shape);
```

Comment:
This isn't safe, check the code above, `node` can be null

Reply:
Ah, my bad! I'm on it

Reply:
> Also it's not guaranteed that this is a physics body so the signal might cause errors if it's not available
Oh. Then I should probably also add a type check?

Reply:
Probably, a cast to the appropriate type `Object::cast_to<PhysicsBody2D/3D>` like above with `Node`

Reply:
Roger /0

Reply:
It would be enough to do:
```cpp
			if (Object::cast_to<PhysicsBody3D>(node)) {
```
As it already handles null, and can still call it on `Node`

Reply:
My first instinct was to do
```gdscript
Object *obj = ObjectDB::get_instance(objid);
Node *node = Object::cast_to<Node>(obj);
PhysicsBody3D *body = Object::cast_to<PhysicsBody3D>(obj);
```

and then

```gdscript
if (body) {
	body->emit_signal(SceneStringNames::get_singleton()->entered_area, this);
}
```
It doesn't have an advantage of being general, and since areas can intersect with other stuff, maybe people using it will want to add signals there too. It's kinda like your first suggestion but with caching the cast :)

Reply:
There's no need to cache the cast, since the call `emit_signal` doesn't change, it's the same regardless what class it's called on, but it's either way, you use your own judgement 🙂

Reply:
I think I'll roll with solution 2 and leave our GridMap fellows the question of arguments :D

Reply:
Holy, this was a lot of `if`s :)

Done